11_FFTBloom/main.cpp

```diff
@@ -0,0 +1,1192 @@
+	{
+		IGPUDescriptorSet::SDescriptorInfo pInfos[CHANNELS];
+		IGPUDescriptorSet::SWriteDescriptorSet write;
+
+		write.dstSet = set;
+		write.binding = 0;
+		write.arrayElement = 0u;
+		write.count = CHANNELS;
+		write.info = pInfos;
+
+		for (uint32_t i = 0u; i < CHANNELS; i++)
+		{
+			auto& info = pInfos[i];
+			info.desc = kernelNormalizedSpectrumImageDescriptors[i];
+			info.info.combinedImageSampler.imageLayout = layout;
+			info.info.combinedImageSampler.sampler = nullptr;
+		}
+
+		m_device->updateDescriptorSets(1, &write, 0u, nullptr);
+	}
+
+	inline void updateDescriptorSetConvolution(IGPUDescriptorSet* set, const smart_refctd_ptr<IGPUImageView>* kernelNormalizedSpectrumImageDescriptors)
+	{
+		updateDescriptorSetConvolutionAndNormalization(set, kernelNormalizedSpectrumImageDescriptors, IImage::LAYOUT::READ_ONLY_OPTIMAL);
+	}
+
+	inline void updateDescriptorSetNormalization(IGPUDescriptorSet* set, const smart_refctd_ptr<IGPUImageView>* kernelNormalizedSpectrumImageDescriptors)
+	{
+		updateDescriptorSetConvolutionAndNormalization(set, kernelNormalizedSpectrumImageDescriptors, IImage::LAYOUT::GENERAL);
+	}
+
+	inline void updateDescriptorSetFirstAxisIFFT(IGPUDescriptorSet* set, smart_refctd_ptr<IGPUImageView> outputImageDescriptor)
+	{
+		IGPUDescriptorSet::SDescriptorInfo info;
+		IGPUDescriptorSet::SWriteDescriptorSet write;
+
+		write.dstSet = set;
+		write.binding = 0;
+		write.arrayElement = 0u;
+		write.count = 1;
+		write.info = &info;
+
+		info.desc = outputImageDescriptor;
+		info.info.combinedImageSampler.imageLayout = IImage::LAYOUT::GENERAL;
+		info.info.combinedImageSampler.sampler = nullptr;
+
+		m_device->updateDescriptorSets(1u, &write, 0u, nullptr);
+	}
+
+	inline core::smart_refctd_ptr<video::IGPUShader> createShader(
+		const char* includeMainName,
+		uint32_t workgroupSize,
+		uint32_t elementsPerThread,
+		float kernelScale = 1.f)
+	{
+
+		const char* sourceFmt =
+			R"===(
+		#define _NBL_HLSL_WORKGROUP_SIZE_ %u
+		#define ELEMENTS_PER_THREAD %u
+		
+		#define KERNEL_SCALE %f
+		#define scalar_t %s
+		#define FORMAT %s
+ 
+		#include "%s"
+
+		)===";
+
+		const size_t extraSize = 4u + 4u + 9u + 7u + 128u;
+
+		auto shader = core::make_smart_refctd_ptr<ICPUBuffer>(strlen(sourceFmt) + extraSize + 1u);
+		snprintf(
+			reinterpret_cast<char*>(shader->getPointer()), shader->getSize(), sourceFmt,
+			workgroupSize,
+			elementsPerThread,
+			kernelScale,
+			useHalfFloats ? "float16_t" : "float32_t",
+			useHalfFloats ? "\"rg16f\"" : "\"rg32f\"",
+			includeMainName
+		);
+
+		auto CPUShader = core::make_smart_refctd_ptr<ICPUShader>(std::move(shader), IShader::E_SHADER_STAGE::ESS_COMPUTE, IShader::E_CONTENT_TYPE::ECT_HLSL, includeMainName);
+		assert(CPUShader);
+
+		ISPIRVOptimizer::E_OPTIMIZER_PASS optPasses = ISPIRVOptimizer::EOP_STRIP_DEBUG_INFO;
```

Comment:
you probably want to gate this with with `ifdef NBL_DEBUG` (check what the define is called though!)